File: sparkJava/src/main/java/com/beifeng/sql/JSONDataSource.java
package com.beifeng.sql;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import org.apache.spark.sql.*;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.List;

/**
 * JSON数据源
 * time: 2018-09-16 11:08
 */
public class JSONDataSource {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf()
                .setAppName("JSONDataSource")
                .setMaster("local");

        JavaSparkContext sc = new JavaSparkContext(conf);
        SQLContext sqlContext = new SQLContext(sc);

        // 针对json文件，创建DataFrame
        DataFrame dfStudentScores = sqlContext.read().json("hdfs://spark1:9000/spark-sql/source/student_scores.json");
        dfStudentScores.registerTempTable("student_scores");
        DataFrame dfGoodStudentScores = sqlContext.sql("select name,score from student_scores where score >= 80");

        List<String> goodStudentNames = dfGoodStudentScores.javaRDD().map(new Function<Row, String>() {
            @Override
            public String call(Row row) throws Exception {
                return row.getString(0);
            }
        }).collect();

        List<String> studentInfoJSONs = new ArrayList<String>();
        studentInfoJSONs.add("{\"name\":\"Leo\", \"age\":18}");
        studentInfoJSONs.add("{\"name\":\"Marry\", \"age\":17}");
        studentInfoJSONs.add("{\"name\":\"Jack\", \"age\":19}");
        JavaRDD<String> rddStudentInfo = sc.parallelize(studentInfoJSONs);
        DataFrame dfStudentInfo = sqlContext.read().json(rddStudentInfo);

        dfStudentInfo.registerTempTable("student_infos");

        String sql = "select name, age from student_infos where name in (";
        for(int i = 0;i < goodStudentNames.size();i ++){
            sql += "'" + goodStudentNames.get(i) + "'";
            if(i < goodStudentNames.size() - 1){
                sql += ",";
            }
        }
        sql += ")";

        DataFrame dfGoodStudentInfo = sqlContext.sql(sql);
        JavaPairRDD<String, Tuple2<Integer, Integer>> goodStudentRDD = dfGoodStudentScores.javaRDD().mapToPair(new PairFunction<Row, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(Row row) throws Exception {
                return new Tuple2<String, Integer>(row.getString(0), (int)row.getLong(1));
            }
        }).join(dfGoodStudentInfo.javaRDD().mapToPair(new PairFunction<Row, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(Row row) throws Exception {
                return new Tuple2<String, Integer>(row.getString(0), (int)row.getLong(1));
            }
        }));

        // 封装在RDD中的好学生的全部信息，转换为一个JavaRDD<Row>的格式
        JavaRDD<Row> goodStudentRowsRDD = goodStudentRDD.map(new Function<Tuple2<String, Tuple2<Integer, Integer>>, Row>() {
            @Override
            public Row call(Tuple2<String, Tuple2<Integer, Integer>> t) throws Exception {
                return RowFactory.create(t._1, t._2._1, t._2._2);
            }
        });

        List<StructField> structFields = new ArrayList<StructField>();
        structFields.add(DataTypes.createStructField("name", DataTypes.StringType, true));
        structFields.add(DataTypes.createStructField("score", DataTypes.IntegerType, true));
        structFields.add(DataTypes.createStructField("age", DataTypes.IntegerType, true));
        StructType structType = DataTypes.createStructType(structFields);
        DataFrame goodStudentsDF = sqlContext.createDataFrame(goodStudentRowsRDD, structType);
        goodStudentsDF.printSchema();
        goodStudentsDF.show();

        goodStudentsDF.save("hdfs://spark1:9000/spark-sql/output/java/good-students", "json", SaveMode.Overwrite);
    }
}

File: sparkJava/src/main/java/com/beifeng/scala/sql/JSONDataSource.scala
package com.beifeng.scala.sql

import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
import org.apache.spark.sql.{Row, SQLContext, SaveMode}
import org.apache.spark.{SparkConf, SparkContext}

object JSONDataSource {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
      .setAppName("JSONDataSource")
      .setMaster("local")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)

    val dfStudentScore = sqlContext.read.json("hdfs://spark1:9000/spark-sql/source/student_scores.json")
    dfStudentScore.registerTempTable("student_scores")
    val listGoodStudentName = sqlContext.sql("select name, score from student_scores where score >= 80").rdd.map(row => row.getString(0))
      .collect()

    val arrStudentInfoJSON = Array(
      "{\"name\":\"Leo\", \"age\":18}",
      "{\"name\":\"Marry\", \"age\":17}",
      "{\"name\":\"Jack\", \"age\":19}"
    )
    val rddStudentInfo = sc.parallelize(arrStudentInfoJSON)
    val dfStudentInfo = sqlContext.read.json(rddStudentInfo)
    dfStudentInfo.registerTempTable("student_infos")

    var sql = "select name,age from student_infos where name in ("
    for (i <- 0 until listGoodStudentName.length) {
      sql += "'" + listGoodStudentName(i) + "'"
      if (i < listGoodStudentName.length - 1) {
        sql += ","
      }
    }
    sql += ")"

    val dfGoodStudentInfos = sqlContext.sql(sql)

    val rddGoodStudentInfosRow = dfGoodStudentInfos.rdd.map ( row =>
      (row.getAs[String]("name"), row.getAs[Long]("age"))
    )

    val rddGoodStudent = dfStudentScore.rdd.map(row => {
      (row.getAs[String]("name"), row.getAs[Long]("score"))
    })
      .join(rddGoodStudentInfosRow)

    val rddGoodStudentRows = rddGoodStudent.map(info => Row(info._1, info._2._1.toInt, info._2._2.toInt))

    val stGoodStudentRow = StructType(Array(
      StructField("name", StringType, true),
      StructField("score", IntegerType, true),
      StructField("age", IntegerType, true)
    ))

    val dfGoodStudents = sqlContext.createDataFrame(rddGoodStudentRows, stGoodStudentRow)
    dfGoodStudents.save("hdfs://spark1:9000/spark-sql/output/scala/good-students", "json", SaveMode.Overwrite)

    sqlContext.read.json("hdfs://spark1:9000/spark-sql/output/scala/good-students")
      .show()
  }
}
